### PR TITLE
Update pythonpublish.yml

### PR DESCRIPTION
**Pull request by kelle**
Updating the workflow to publish to PyPI following this guide: https://packaging.python.org/en/latest/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows/#workflow-definition

### CODE REVIEW

**Review by dr-rodriguez — APPROVED** [review]
Not super familiar with all the setup, but if this is what's recommended then that's fine with me.

**Comment by dr-rodriguez** [comment]
Do we want to delete the 0.2 tag after we merge this? The 0.2 is NOT using this mechanism since its behind

File: .github/workflows/pythonpublish.yml [hunk 1]
```diff
@@ -4,6 +4,8 @@
 name: Publish Python Package to PyPI
 
 on:
+  pull_request:     # delete this once workflow tested
+    branches: [main]  # delete this once workflow tested
```
Comment (dr-rodriguez) [inline .github/workflows/pythonpublish.yml:8]: We also may have wanted to switch to using testpypi for testing the workflow

Comment (kelle) [reply .github/workflows/pythonpublish.yml:8]: ok, I'll do that.

File: .github/workflows/pythonpublish.yml [hunk 2]
```diff
@@ -1,32 +1,58 @@
...
+    - name: Store the distribution package
+      uses: actions/upload-artifact@v4
+      with:
+        name: python-package-distributions
+        path: dist/
+        
+  publish-to-pypi:
+    name: >-
+      Publish to PyPI
+    if: startsWith(github.ref, 'refs/tags/')  # only publish to PyPI on tag pushes
```
Comment (kelle) [inline .github/workflows/pythonpublish.yml:42]: @dr-rodriguez , I don't understand this bit.  If we trigger on release, do we also need this?

**Comment by kelle** [comment]
> Do we want to delete the 0.2 tag after we merge this? The 0.2 is NOT using this mechanism since its behind

My plan was to trigger the workflow manually once this workflow is updated.  But if that doesn't work, then yes, we need to make a new release. I'm ok with it being a new 0.2.1 rather than deleting 0.2.

File: .github/workflows/pythonpublish.yml (hunk 2 above)
Comment (dr-rodriguez) [reply .github/workflows/pythonpublish.yml:66]: A release creates a tag, so this is just checking that. We do need it, I think.

File: .github/workflows/pythonpublish.yml (hunk 1 above)
Comment (kelle) [reply .github/workflows/pythonpublish.yml:8]: I've created an account on testpypi but the package isn't there for me to manage. How do I get it on testpypi for the very first time?

Comment (kelle) [reply .github/workflows/pythonpublish.yml:8]: Found it, doing it. https://packaging.python.org/en/latest/tutorials/packaging-projects/#uploading-the-distribution-archives

**Comment by dr-rodriguez** [comment]
> > Do we want to delete the 0.2 tag after we merge this? The 0.2 is NOT using this mechanism since its behind
> 
> My plan was to trigger the workflow manually once this workflow is updated. But if that doesn't work, then yes, we need to make a new release. I'm ok with it being a new 0.2.1 rather than deleting 0.2.

The original workflow doesn't allow for manual execution (I think that's what `worfklow_dispatch` enables). But if you're OK with just pushing 0.2.1 then that's fine with me.

File: .github/workflows/pythonpublish.yml (hunk 1 above)
Comment (kelle) [reply .github/workflows/pythonpublish.yml:8]: waiting for Arjun to give me permissions on the package in testpypi.

**Comment by kelle** [comment]
I'm stuck. 
```
400 The use of local versions in                              
         <Version('0.1.dev1+g4a4cdb0')> is not allowed. See                     
         https://packaging.python.org/specifications/core-metadata for more     
         information.</title>                   
```
Going to lunch.

**Comment by kelle** [comment]
Abandoned the testpypi recommended workflow since we (me and Lehman) couldn't figure out how to make it work with dynamic versioning.  This workflow works the same way the old one did, just with the newer actions and authentication mechanism.

**Pull request merged**